pymongo/client_bulk_shared.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2024-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you
+# may not use this file except in compliance with the License.  You
+# may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.  See the License for the specific language governing
+# permissions and limitations under the License.
+
+
+"""Constants, types, and classes shared across Client Bulk Write API implementations."""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Mapping, MutableMapping, NoReturn, Optional
+
+from pymongo.errors import OperationFailure, WriteConcernError, WriteError
+from pymongo.helpers_shared import _get_wce_doc
+from pymongo.results import ClientBulkWriteResult
+
+if TYPE_CHECKING:
+    from pymongo.typings import _DocumentOut
+
+
+class ClientBulkWriteException(OperationFailure):
```

Comment:
What do you think about moving this errors.py where all the other exceptions live?

Reply:
When I put it in errors.py I get a circular import error with results.py -- I think because I import `ClientBulkWriteResult` from results.py for the `partial_result` field of the exception

Reply:
We could make it a delayed import:
```
    @property
    def partial_result(self) -> Optional[ClientBulkWriteResult]:
        from pymongo.results import ClientBulkWriteResult
        ...
```

Reply:
Got it I'll move it to errors.py!